x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,680 @@
+}
+
+func (k Keeper) GetLVG(ctx sdk.Context, bucketID math.Uint, lvgID uint32) (*types.LocalVirtualGroup, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetLVGKey(bucketID, lvgID))
+	if bz == nil {
+		return nil, false
+	}
+	var lvg types.LocalVirtualGroup
+	k.cdc.MustUnmarshal(bz, &lvg)
+	return &lvg, true
+}
+
+func (k Keeper) GetLVGs(ctx sdk.Context, bucketID math.Uint) []*types.LocalVirtualGroup {
+	lvgs := make([]*types.LocalVirtualGroup, 0)
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, types.GetLVGKey(bucketID, 0))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var lvg types.LocalVirtualGroup
+		k.cdc.MustUnmarshal(iterator.Value(), &lvg)
+		lvgs = append(lvgs, &lvg)
+	}
+
+	return lvgs
+}
+
+func (k Keeper) GetGVGsBindingOnBucket(ctx sdk.Context, bucketID math.Uint) (*types.GlobalVirtualGroupsBindingOnBucket, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.GetGVGsBindingOnBucketKey(bucketID))
+	if bz == nil {
+		return nil, false
+	}
+
+	var gvgsBindingOnBucket types.GlobalVirtualGroupsBindingOnBucket
+	k.cdc.MustUnmarshal(bz, &gvgsBindingOnBucket)
+	return &gvgsBindingOnBucket, true
+}
+
+func (k Keeper) SetGVGsBindingOnBucket(ctx sdk.Context, gvgsBindingOnBucket *types.GlobalVirtualGroupsBindingOnBucket) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsBindingOnBucket)
+	store.Set(types.GetGVGsBindingOnBucketKey(gvgsBindingOnBucket.BucketId), bz)
+}
+
+func (k Keeper) SetGVGFamily(ctx sdk.Context, primarySpID uint32, gvgFamily *types.GlobalVirtualGroupFamily) {
+
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgFamily)
+	store.Set(types.GetGVGFamilyKey(primarySpID, gvgFamily.Id), bz)
+}
+
+func (k Keeper) GetGVGFamily(ctx sdk.Context, spID, familyID uint32) (*types.GlobalVirtualGroupFamily, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	var gvgFamily types.GlobalVirtualGroupFamily
+	bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+	if bz == nil {
+		return nil, false
+	}
+	k.cdc.MustUnmarshal(bz, &gvgFamily)
+	return &gvgFamily, true
+}
+
+func (k Keeper) GenerateOrSetLVGForBucket(ctx sdk.Context, bucketID math.Uint, gvgID uint32) {
+}
+
+func (k Keeper) GetOrCreateEmptyGVGFamily(ctx sdk.Context, familyID uint32, spID uint32) (*types.GlobalVirtualGroupFamily, error) {
+	store := ctx.KVStore(k.storeKey)
+	var gvgFamily types.GlobalVirtualGroupFamily
+	if familyID == types.NoSpecifiedFamilyId {
+		id := k.GenNextGVGFamilyID(ctx)
+		gvgFamily = types.GlobalVirtualGroupFamily{
+			Id:                    id,
+			VirtualPaymentAddress: k.DeriveVirtualPaymentAccount(types.GVGFamilyName, id).String(),
+		}
+		return &gvgFamily, nil
+	} else {
+		bz := store.Get(types.GetGVGFamilyKey(spID, familyID))
+		if bz == nil {
+			return nil, types.ErrGVGFamilyNotExist
+		}
+		k.cdc.MustUnmarshal(bz, &gvgFamily)
+
+		storeSize := k.GetStoreSizeOfFamily(ctx, gvgFamily)
```

Comment:
Do we need checks in line 252-257? Here we just get the gvg family, no updates.

Reply:
Yep. Move these to `func (k msgServer) CreateGlobalVirtualGroup(goCtx context.Context, req *types.MsgCreateGlobalVirtualGroup) (*types.MsgCreateGlobalVirtualGroupResponse, error`